File: playwright_smoke/demo-server.js
const express = require('express');
const path = require('path');
const app = express();
const PORT = 3000;

// Middleware
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// In-memory storage
let products = [
    { id: 1, name: 'Laptop', price: 999, stock: 10 },
    { id: 2, name: 'Phone', price: 599, stock: 5 },
    { id: 3, name: 'Tablet', price: 399, stock: 8 }
];

let users = [];
let orders = [];
let cart = [];

// HTML template function
function htmlTemplate(title, body) {
    return `
<!DOCTYPE html>
<html>
<head>
    <title>${title}</title>
    <style>
        body { font-family: Arial, sans-serif; margin: 40px; }
        .nav { background: #333; padding: 10px; margin-bottom: 20px; }
        .nav a { color: white; text-decoration: none; margin-right: 20px; }
        .product { border: 1px solid #ddd; padding: 10px; margin: 10px 0; }
        .btn { background: #007bff; color: white; padding: 8px 16px; border: none; cursor: pointer; }
        .btn:hover { background: #0056b3; }
        .error { color: red; }
        .success { color: green; }
        .cart-item { background: #f8f9fa; padding: 10px; margin: 5px 0; }
    </style>
</head>
<body>
    <div class="nav">
        <a href="/">Home</a>
        <a href="/products">Products</a>
        <a href="/cart">Cart (${cart.length})</a>
        <a href="/login">Login</a>
        <a href="/register">Register</a>
    </div>
    ${body}
</body>
</html>`;
}

// Routes
app.get('/', (req, res) => {
    const body = `
    <h1>Welcome to Demo E-Commerce</h1>
    <p>This is a demo e-commerce site for smoke testing.</p>
    <p><a href="/products" class="btn">Browse Products</a></p>
    <p><a href="/cart" class="btn">View Cart (${cart.length} items)</a></p>
  `;
    res.send(htmlTemplate('Home - Demo Shop', body));
});

app.get('/products', (req, res) => {
    const productList = products.map(p => `
    <div class="product">
      <h3>${p.name}</h3>
      <p>Price: $${p.price}</p>
      <p>Stock: ${p.stock}</p>
      <form method="post" action="/cart/add" style="display: inline;">
        <input type="hidden" name="productId" value="${p.id}">
        <button type="submit" class="btn">Add to Cart</button>
      </form>
    </div>
  `).join('');

    const body = `
    <h1>Products</h1>
    ${productList}
  `;
    res.send(htmlTemplate('Products - Demo Shop', body));
});

app.get('/cart', (req, res) => {
    const cartItems = cart.map(item => {
        const product = products.find(p => p.id === item.productId);
        return `
      <div class="cart-item">
        <h4>${product.name}</h4>
        <p>Price: $${product.price}</p>
        <p>Quantity: ${item.quantity}</p>
        <form method="post" action="/cart/remove" style="display: inline;">
          <input type="hidden" name="productId" value="${item.productId}">
          <button type="submit" class="btn">Remove</button>
        </form>
      </div>
    `;
    }).join('');

    const total = cart.reduce((sum, item) => {
        const product = products.find(p => p.id === item.productId);
        return sum + (product.price * item.quantity);
    }, 0);

    const body = `
    <h1>Shopping Cart</h1>
    ${cartItems || '<p>Your cart is empty</p>'}
    <h3>Total: $${total}</h3>
    ${cart.length > 0 ? '<a href="/checkout" class="btn">Checkout</a>' : ''}
  `;
    res.send(htmlTemplate('Cart - Demo Shop', body));
});

app.post('/cart/add', (req, res) => {
    const productId = parseInt(req.body.productId);
    const existingItem = cart.find(item => item.productId === productId);

    if (existingItem) {
        existingItem.quantity += 1;
    } else {
        cart.push({ productId, quantity: 1 });
    }

    res.redirect('/cart');
});

app.post('/cart/remove', (req, res) => {
    const productId = parseInt(req.body.productId);
    cart = cart.filter(item => item.productId !== productId);
    res.redirect('/cart');
});

app.get('/login', (req, res) => {
    const body = `
    <h1>Login</h1>
    <form method="post" action="/login">
      <p>
        <label>Username:</label><br>
        <input type="text" name="username" required>
      </p>
      <p>
        <label>Password:</label><br>
        <input type="password" name="password" required>
      </p>
      <p>
        <button type="submit" class="btn">Login</button>
      </p>
    </form>
    <p><a href="/register">Don't have an account? Register here</a></p>
  `;
    res.send(htmlTemplate('Login - Demo Shop', body));
});

app.post('/login', (req, res) => {
    const { username, password } = req.body;
    const user = users.find(u => u.username === username && u.password === password);

    if (user) {
        res.redirect('/');
    } else {
        const body = `
      <h1>Login</h1>
      <p class="error">Invalid username or password</p>
      <form method="post" action="/login">
        <p>
          <label>Username:</label><br>
          <input type="text" name="username" required>
        </p>
        <p>
          <label>Password:</label><br>
          <input type="password" name="password" required>
        </p>
        <p>
          <button type="submit" class="btn">Login</button>
        </p>
      </form>
    `;
        res.send(htmlTemplate('Login - Demo Shop', body));
    }
});

app.get('/register', (req, res) => {
    const body = `
    <h1>Register</h1>
    <form method="post" action="/register">
      <p>
        <label>Username:</label><br>
        <input type="text" name="username" required>
      </p>
      <p>
        <label>Email:</label><br>
        <input type="email" name="email" required>
      </p>
      <p>
        <label>Password:</label><br>
        <input type="password" name="password" required>
      </p>
      <p>
        <button type="submit" class="btn">Register</button>
      </p>
    </form>
    <p><a href="/login">Already have an account? Login here</a></p>
  `;
    res.send(htmlTemplate('Register - Demo Shop', body));
});

app.post('/register', (req, res) => {
    const { username, email, password } = req.body;

    if (users.find(u => u.username === username)) {
        const body = `
      <h1>Register</h1>
      <p class="error">Username already exists</p>
      <form method="post" action="/register">
        <p>
          <label>Username:</label><br>
          <input type="text" name="username" required>
        </p>
        <p>
          <label>Email:</label><br>
          <input type="email" name="email" required>
        </p>
        <p>
          <label>Password:</label><br>
          <input type="password" name="password" required>
        </p>
        <p>
          <button type="submit" class="btn">Register</button>
        </p>
      </form>
    `;
        res.send(htmlTemplate('Register - Demo Shop', body));
    } else {
        users.push({ username, email, password });
        res.redirect('/login');
    }
});

app.get('/checkout', (req, res) => {
    const total = cart.reduce((sum, item) => {
        const product = products.find(p => p.id === item.productId);
        return sum + (product.price * item.quantity);
    }, 0);

    const body = `
    <h1>Checkout</h1>
    <h3>Order Total: $${total}</h3>
    <form method="post" action="/checkout">
      <p>
        <label>Card Number:</label><br>
        <input type="text" name="cardNumber" placeholder="1234 5678 9012 3456" required>
      </p>
      <p>
        <label>Expiry Date:</label><br>
        <input type="text" name="expiry" placeholder="MM/YY" required>
      </p>
      <p>
        <label>CVV:</label><br>
        <input type="text" name="cvv" placeholder="123" required>
      </p>
      <p>
        <button type="submit" class="btn">Place Order</button>
        <a href="/cart" class="btn" style="background: #6c757d;">Back to Cart</a>
      </p>
    </form>
  `;
    res.send(htmlTemplate('Checkout - Demo Shop', body));
});

app.post('/checkout', (req, res) => {
    const orderId = orders.length + 1;
    orders.push({
        id: orderId,
        items: [...cart],
        timestamp: new Date()
    });
    cart = []; // Clear cart

    const body = `
    <h1>Order Confirmation</h1>
    <p class="success">Thank you! Your order #${orderId} has been placed successfully.</p>
    <p><a href="/" class="btn">Continue Shopping</a></p>
  `;
    res.send(htmlTemplate('Order Confirmation - Demo Shop', body));
});

// API Routes
app.get('/api/products', (req, res) => {
    res.json(products);
});

app.get('/api/health', (req, res) => {
    res.json({ status: 'ok', timestamp: new Date().toISOString() });
});

// 404 handler
app.use((req, res) => {
    res.status(404).send(htmlTemplate('Page Not Found', '<h1>404 - Page Not Found</h1>'));
});

app.listen(PORT, () => {
    console.log(`🚀 Demo E-Commerce Server running at http://localhost:${PORT}`);
    console.log(`📝 Demo data initialized with ${products.length} products`);
    console.log(`🛑 Press Ctrl+C to stop the server`);
});

File: demo_server.py
#!/usr/bin/env python3
"""
Demo Flask Server for Smoke Testing
Mô phỏng một ứng dụng Microblog đơn giản để test
"""

from flask import Flask, render_template_string, request, redirect, url_for, session, jsonify
import os
import json

app = Flask(__name__)
app.secret_key = 'demo_secret_key_for_testing'

# In-memory storage for demo
users = {}
posts = []

# HTML Templates
HOME_TEMPLATE = '''
<!DOCTYPE html>
<html>
<head><title>Demo Microblog</title></head>
<body>
    <h1>Welcome to Demo Microblog</h1>
    {% if session.get('username') %}
        <p>Hello, {{ session.username }}!</p>
        <form method="POST" action="/">
            <textarea name="post" placeholder="What's on your mind?" required></textarea><br>
            <button type="submit" name="submit">Submit</button>
        </form>
        <p><a href="/auth/logout">Logout</a></p>
    {% else %}
        <p><a href="/auth/login">Login</a> | <a href="/auth/register">Register</a></p>
    {% endif %}
    
    <h2>Recent Posts</h2>
    {% for post in posts %}
        <div style="border:1px solid #ccc; margin:10px; padding:10px;">
            <strong>{{ post.username }}:</strong> {{ post.content }}
        </div>
    {% endfor %}
    
    <p><a href="/explore">Explore</a></p>
</body>
</html>
'''

LOGIN_TEMPLATE = '''
<!DOCTYPE html>
<html>
<head><title>Login - Demo Microblog</title></head>
<body>
    <h1>Login</h1>
    <form method="POST">
        <input type="text" name="username" placeholder="Username" required><br><br>
        <input type="password" name="password" placeholder="Password" required><br><br>
        <input type="checkbox" name="remember_me"> Remember me<br><br>
        <button type="submit" name="submit">Sign In</button>
    </form>
    <p><a href="/auth/register">Don't have an account? Register here</a></p>
    <p><a href="/">Back to Home</a></p>
</body>
</html>
'''

REGISTER_TEMPLATE = '''
<!DOCTYPE html>
<html>
<head><title>Register - Demo Microblog</title></head>
<body>
    <h1>Register</h1>
    <form method="POST">
        <input type="text" name="username" placeholder="Username" required><br><br>
        <input type="email" name="email" placeholder="Email" required><br><br>
        <input type="password" name="password" placeholder="Password" required><br><br>
        <input type="password" name="password2" placeholder="Repeat Password" required><br><br>
        <button type="submit" name="submit">Register</button>
    </form>
    <p><a href="/auth/login">Already have an account? Login here</a></p>
    <p><a href="/">Back to Home</a></p>
</body>
</html>
'''

EXPLORE_TEMPLATE = '''
<!DOCTYPE html>
<html>
<head><title>Explore - Demo Microblog</title></head>
<body>
    <h1>Explore Posts</h1>
    {% for post in posts %}
        <div style="border:1px solid #ccc; margin:10px; padding:10px;">
            <strong>{{ post.username }}:</strong> {{ post.content }}
        </div>
    {% endfor %}
    <p><a href="/">Back to Home</a></p>
</body>
</html>
'''

@app.route('/', methods=['GET', 'POST'])
def index():
    if request.method == 'POST':
        if session.get('username'):
            content = request.form.get('post')
            if content:
                posts.append({
                    'username': session['username'],
                    'content': content
                })
        return redirect(url_for('index'))
    
    return render_template_string(HOME_TEMPLATE, posts=posts[:10])

@app.route('/index')
def index_alt():
    return redirect(url_for('index'))

@app.route('/auth/login', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        username = request.form.get('username')
        password = request.form.get('password')
        
        if username in users and users[username]['password'] == password:
            session['username'] = username
            return redirect(url_for('index'))
        else:
            return render_template_string(LOGIN_TEMPLATE + '<p style="color:red;">Invalid credentials</p>')
    
    return render_template_string(LOGIN_TEMPLATE)

@app.route('/auth/register', methods=['GET', 'POST'])
def register():
    if request.method == 'POST':
        username = request.form.get('username')
        email = request.form.get('email')
        password = request.form.get('password')
        password2 = request.form.get('password2')
        
        if password != password2:
            return render_template_string(REGISTER_TEMPLATE + '<p style="color:red;">Passwords do not match</p>')
        
        if username in users:
            return render_template_string(REGISTER_TEMPLATE + '<p style="color:red;">Username already exists</p>')
        
        users[username] = {
            'email': email,
            'password': password
        }
        
        return redirect(url_for('login'))
    
    return render_template_string(REGISTER_TEMPLATE)

@app.route('/auth/logout')
def logout():
    session.pop('username', None)
    return redirect(url_for('index'))

@app.route('/explore')
def explore():
    return render_template_string(EXPLORE_TEMPLATE, posts=posts)

# API Endpoints
@app.route('/api')
def api_root():
    return jsonify({
        'message': 'Demo Microblog API',
        'version': '1.0',
        'endpoints': ['/api/users', '/api/tokens']
    })

@app.route('/api/users')
def api_users():
    # Simulate authentication requirement
    if 'Authorization' in request.headers:
        return jsonify({
            'users': [{'username': user, 'email': data['email']} for user, data in users.items()]
        })
    else:
        return jsonify({'error': 'Authentication required'}), 401

@app.route('/api/tokens', methods=['POST'])
def api_tokens():
    # Simulate token endpoint
    if not request.is_json:
        return jsonify({'error': 'Content-Type must be application/json'}), 400
    
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Invalid JSON'}), 400
        
    username = data.get('username')
    password = data.get('password')
    
    if username and password and username in users and users[username]['password'] == password:
        return jsonify({'token': f'demo_token_{username}'})
    else:
        return jsonify({'error': 'Invalid credentials'}), 401

@app.route('/api/tokens', methods=['GET'])
def api_tokens_get():
    return jsonify({'error': 'Method not allowed'}), 405

@app.route('/api/nonexistent')
def api_nonexistent():
    return jsonify({'error': 'Endpoint not found'}), 404

if __name__ == '__main__':
    print("🚀 Starting Demo Microblog Server...")
    print("📝 Default test user will be created: testuser/password123")
    
    # Create a default test user
    users['testuser'] = {
        'email': 'testuser@example.com',
        'password': 'password123'
    }
    
    # Add some sample posts
    posts.append({
        'username': 'testuser',
        'content': 'Welcome to the demo microblog!'
    })
    
    print("🌐 Server will be available at: http://localhost:5000")
    print("🛑 Press Ctrl+C to stop the server")
    
    app.run(host='localhost', port=5000, debug=True, use_reloader=False)
